agreement/asyncVoteVerifier.go

```diff
@@ -174,7 +180,9 @@ func (avv *AsyncVoteVerifier) verifyEqVote(verctx context.Context, l LedgerReade
 // Quit tells the AsyncVoteVerifier to shutdown and waits until all workers terminate.
 func (avv *AsyncVoteVerifier) Quit() {
 	// indicate we're done and wait for all workers to finish
+	avv.mu.Lock()
 	avv.ctxCancel()
+	avv.mu.Unlock()
```

Comment:
No, they are not equivalent. 

Alternative solution is the following:
Instead of using the wg (which tracks the number of tasks submitted in order to wait for them before closing), we can use a channel or an atomic variable which will keep track of the submitted tasks. Then, by replacing the wg.Add() and moving it to before the ctx check, we guarantee that the shutdown will still wait for the task (submitted after the ctx was canceled), before it closes the channel. 

The problem with using a channel is that the size will grow with the number of elements in the pipeline. 
The problem with using an atomic variable is that we still need to add synchronization around it. Synchronization is not guaranteed by the atomic variable.

Reply:
`close(avv.execpppoolOut)` is preceded by `avv.wg.Wait()`.  So, as long as the `wg.Add()` is called before `avv.ctxCancel()`, `excecpoolOut` will not close before it comes out (`avv.wg.Wait()`). 

> I thought the fix would involve pool.worker() checking some kind of atomic/channel/lock to ensure that it is safe to send a result to the output channel..?
It is the same problem over there. You need to check then send, and make sure nothing changes between checking and sending. So, mutex?

Reply:
The channel that the exec pool writes to is owned by the application which is using the exec pool. The exec pool should not worry about it. The exec pool is given a channel to return the results to, so whoever gave it should make sure the channel is open by the time the results are returned. 

In our case, `avv.wg` counts the number of tasks sent to the exec pool, and decrements `avv.wg` each time a results comes out of the channel. When `avv.wg` is 0, it means no more tasks will come out of the channel. After that, the out channel is closed (after `avv.wg.Wait()`).

Here are the interaction of 3 goroutines: A is `AsyncVoteVerifier.Quit`, B is `AsyncVoteVerifier.veifyVote`, C is the execpool returning the result. The scheduling of each step is given in square brackets. 

The bug was a synchronization bug. It goes like this: 
A: =========== [2]`ctx.Cancel` =[3]`wg.Wait`= [4/5]`close(outchan)`
B: [1]`ctx.Done?`========================[4/5]`wg.Add(1)`=[6]`enQueue(vote)`
C: ===============================================================[7]`outchan<-result`

With the fix, you cannot have `ctx.Cancel` come in between `ctx.Done?` and `wg.Add(1)`, and consequently, cannot have `wg.Wait` come before `wg.Add(1)`.